### ISSUE TITLE
BUG: jax2tf (enable_xla=False): Incorrect results for reduce_window, lax.add and padding="SAME"

### ISSUE THREAD

**Issue by marcvanzee** [google/jax#11874]
### Description

In jax2tf (`enable_xla=False`) we use `tf.nn.avg_pool` to implement `lax.reduce_window_sum_p`. This is because TF does not have a "sum pooling" layer. The idea is that we take the average over a window and then multiply with the window size. Suppose the following input

```
[[1, 2]
 [3, 4]]
```

And suppose further that `window_size=(2, 2)`, `strides=(1,1)`, and `padding="VALID"`.
* `lax.reduce_window(lax.sum)` returns `10`.
* `tf.nn.avg_pool` returns `10/4 = 2.5.` Then `2.5*prod((2, 2)) = 2.5*4 = 10`.

This seems to work fine, but unfortunately it breaks when using `SAME` padding with `window_size != (1, 1, ..)`.
The problem is that `tf.nn.avg_pool` returns the average ignoring the padding tokens, so multiplying with `np.prod(window)` will give too high outputs.

## Example Failure

Here is an example: Suppose the same input and `window_size` as before, but now with `SAME` padding. This means the input will first be padded as follows:
```
[[1, 2, 0]
 [3, 4, 0]
 [0, 0, 0]]
```

Now `lax.reduce_window(lax.sum)` returns:
```
[[10, 6]
 [ 7, 4]]
```

But `tf.nn.avg_pool` returns:
```
[[2.5, 3]
 [3.5, 4]]
```

As we can see, the averaging happens only over the non-padding tokens! The average `3` is computed by doing `(2+4)/2`, rather than `(2+4)/4`. As a result, when we now multiply this output with `np.prod(window_size)`, we get:

```
[[10, 12]
 [14, 16]]
```

This is wrong! 

Unfortunately TF does not allow us to choose whether to include or exclude padding tokens. This is set in the [source code](https://github.com/tensorflow/tensorflow/blob/9e9409a4e32b8653a3f901e4db1cec0ce9941786/tensorflow/stream_executor/cuda/cuda_dnn.cc#L669) but is not exposed to the user.

## Solution

One way to resolve this is to manually pad the input when we are doing SAME padding. 

In our example, this would mean the following:

```
x_pad = tf.pad(x, tf.constant([[0, 0], [0, 1], [0, 1], [0, 0]]))
tf.nn.avg_pool(x_pad, [2, 2], [1, 1], "VALID")
```

This then gives the correct output.

One thing to consider is that `lax.pads_to_padtype` has a different padding algorithm than TF, so we should probably implement it manually (explained [here](https://www.tensorflow.org/api_docs/python/tf/nn#notes_on_padding_2)).

**Comment by sdenton4** [comment]
Thinking about this further, I don't think the TF approach is wrong, and is preferable to the Jax implementation.

Imagine we have a continuous signal, and in order to estimate its moving mean **m**, we sample it and compute an avg-pool operation on the sampled version. Including the padding in the mean estimate will bias the estimate of **m** towards zero. Omitting the padding will lead to an unbiased (but noisier) estimate of **m**. 

In general, the padding is an implementation detail included to ease implementation of the pooling op. If we include it in the pooling operation, we're letting the implementation details impact our computations, which is sad panda.

**Comment by marcvanzee** [comment]
>  I don't think the TF approach is wrong

Yes, I agree! I think including padding tokes in computing the average is a bit meaningless.
I don't mean to say here TF is wrong, but rather that we should make sure we are aware of this discrepancy since it will lead to different results when using jax2tf.

**Comment by sdenton4** [comment]
Yah, I'm thinking the best strategy here is raising this as a bug with the Jax/Linen pooling operation, and thus getting Jax to match the TensorFlow approach.

**Comment by hawkinsp** [comment]
Why is this a bug in JAX/Linen? Isn't this a bug strictly in `jax2tf` where it fails to correctly implement the JAX sum-pooling semantics?

**Comment by sdenton4** [comment]
Per my previous comment: Incorporating the padding values into the pooling operation is bad behavior.

Flax handles the min/max pool correctly by using +/- jnp.inf as 'init' values for the reduction operation; thus, the padding never has an effect on the actual output values. Otherwise, if you took the min_pool over a bunch of strictly positive data, you would get zeros at the edges, instead of useful info about your array.

In the case of average padding, Flax computes sum_pool(x) / window_size. But in the presence of zero-padding the window_size needs to be a bit smaller at the edges to reflect the actual number of non-padding data values being averaged.

**Comment by hawkinsp** [comment]
Correct, if Flax is doing that, then it's wrong. It should be the ratio of two `sum_pool` in that case, see how `tf2xla` does it:
https://source.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/xla/client/lib/pooling.cc;rcl=467457464;l=109

The same translation should work for `lax`.

**Comment by sdenton4** [comment]
Here's the Flax implementation of avg_pool, with the division by np.prod(window_size):
https://github.com/google/flax/blob/main/flax/linen/pooling.py#L87

**Comment by marcvanzee** [comment]
Just to be clear, I think there are two separate issues here:
* Implementing `lax.reduce_window` with computation `lax.add` for jax2tf with `enable_xla=False`. That is what this issue is about: we implement this using `tf.nn.avg_pool`, but this gives wrong results when using SAME padding. I describe this in the issue description. 
* Implementing avg_pool correctly in Flax. Flax is currently counting the padding tokens when doing avg_pool, which I agree seems wrong, but it independent of this issue and should be fixed/discussing in the Flax repo. I will raise a separate issue for this.
